### PR TITLE
fix(cli): validate --max-turns as a positive integer

### PR DESCRIPTION
**Pull request by 0xfandom**
### Problem

`--max-turns` is documented as a safety/cost limit: *"early exit the conversation after the specified number of turns."* It was registered with a bare `.argParser(Number)`, so any value `Number()` coerces was stored verbatim — no NaN, integer, or positivity check.

The sole enforcement point, `query.ts`:

```js
if (maxTurns && nextTurnCount > maxTurns && !nextAgentStepLimit?.summaryRequested) {
  // early exit
}
```

is `falsy` when `maxTurns` is `NaN` or `0`, so the cap never fires. Demonstrated over a 1000-turn run:

```
--max-turns "abc"    Number=NaN       cap fires: NEVER (unbounded)
--max-turns "10x"    Number=NaN       cap fires: NEVER (unbounded)
--max-turns "0"      Number=0         cap fires: NEVER (unbounded)
--max-turns "1e999"  Number=Infinity  cap fires: NEVER (unbounded)
--max-turns "-5"     Number=-5        cap fires: turn 1 (immediate exit)
--max-turns "2.5"    Number=2.5       accepted (non-integer)
--max-turns "3"      Number=3         cap fires: turn 4 (correct)
```

So a typo like `claude -p --max-turns 10x "…"` silently turns the turn cap into **no cap at all** and lets the agent run unbounded — exactly where `--print`/SDK automation relies on it to bound spend.

### Fix

Parse through a `parsePositiveIntArg` helper that rejects NaN, zero, negatives, non-integers, and Infinity at parse time — the same validation the immediately-adjacent `--task-budget` parser already performs. Helper lives in a small unit-tested util.

Note: the neighbouring deprecated `--max-thinking-tokens` uses the same bare `.argParser(Number)`; left out of this change to keep it focused on the documented safety limit, happy to fold it in if preferred.

## Summary by CodeRabbit

- **New Features**
  - Centralized validation for positive numeric CLI options, improving consistency across `--max-turns`, `--task-budget`, and `--max-budget-usd`.

- **Bug Fixes**
  - Enforced stricter rules: only positive safe integers for turn/budget counts and positive, finite, precisely representable amounts for `--max-budget-usd`.
  - More robust rejection of unsafe/extreme values (including problematic scientific notation), with clearer errors that name the failing option.

- **Tests**
  - Expanded test coverage for valid boundary values, extensive invalid inputs, and precision-focused decimal scenarios.

### CODE REVIEW

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Reject unsafe turn-limit integers
  `src/utils/cli/numericArgs.ts:16`
  `Number.isInteger` accepts values outside JavaScript's safe-integer range. For example, `--max-turns 9007199254740993` is accepted but parsed as `9007199254740992`; values such as `1e308` are accepted too. The query loop uses that parsed number as its turn/cost boundary, so the requested cap can be silently changed or become practically unreachable. Require `Number.isSafeInteger(parsed)` (as the adjacent frontmatter max-turn parser does) and cover an unsafe-integer input in the regression tests.

**Comment by 0xfandom** [comment]
Rebased onto current main. Switched the bound to `Number.isSafeInteger` (matching the `maxSteps` check in runAgent.ts) — `9007199254740993` was silently becoming `...992` and `1e308` passed as an unreachable cap. Also routed `--task-budget` through the same helper so both options validate identically and surface Commander's `InvalidArgumentError`. Added regressions for the unsafe-integer inputs and the boundary value.

**Comment by 0xfandom** [comment]
Fixed — `--max-budget-usd` now throws `InvalidArgumentError` like the neighbouring numeric options, so Commander prints the usual validation message instead of a stack trace.

File: src/main.tsx [hunk 1]
```diff
@@ -945,19 +946,13 @@ async function run(): Promise<CommanderCommand> {
...
     const amount = Number(value);
     if (isNaN(amount) || amount <= 0) {
-      throw new Error('--max-budget-usd must be a positive number greater than 0');
+      throw new InvalidArgumentError('--max-budget-usd must be a positive number greater than 0');
```
[the remark this answers is not in the document]
Comment (0xfandom) [reply src/main.tsx]: Added. The parser was inline in the option registration, so I extracted it as `parsePositiveAmountArg` next to `parsePositiveIntArg` and tested it directly — accepted amounts (including decimals), non-numeric/empty/zero/negative, and the error message text.

One thing the tests surfaced while writing them: the original check was `isNaN(amount) || amount <= 0`, and `Number('Infinity')` is not NaN and is greater than zero — so `--max-budget-usd Infinity` (or `1e999`) passed validation as a spending cap that can never be reached. The extracted version uses `Number.isFinite`, so those now reject. Flagging it since it is a small behavior change beyond the error-type fix.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P1] Restore the failing Node 24 smoke-and-tests check
  `smoke-and-tests (24.11.x)`
  The current head fails this required CI job in its “Smoke and full unit test suite” step, so the PR is not merge-ready. Please investigate and fix or otherwise resolve the failing check before merge.

- [P2] Reject finite but unenforceable budget caps
  `src/utils/cli/numericArgs.ts:35`
  `Number.isFinite` still accepts inputs such as `1e308` (and silently rounds `9007199254740993`). That value is forwarded unchanged to the `getTotalCost() >= maxBudgetUsd` guard, so it makes `--max-budget-usd` effectively unbounded—the same failure mode this extraction is intended to prevent—while reporting a different cap from the one the user supplied. Bound this option to an enforceable representable range (or parse a validated monetary representation) and add regressions for the large finite and precision-losing cases.

**Comment by 0xfandom** [comment]
Both addressed. The Node 24 failure was `knowledgeGraph.stress.test.ts` > high-volume entity insertion timing out at 5000ms (took 8216ms) — an unrelated file with no connection to this change; the rebase re-runs it.

On the budget cap: bounded to Number.MAX_SAFE_INTEGER. 1e308 is finite but `getTotalCost() >= maxBudgetUsd` can never fire, and past the safe range the enforced cap isn't the number the user typed. Regressions for both.

**Review by jatmn — APPROVED** [review]
@kevincodex1 LGTM

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Reject precision-losing decimal budget caps
  `src/utils/cli/numericArgs.ts:48`
  The `MAX_SAFE_INTEGER` check only protects integer values. Decimal literals at this magnitude are still accepted after `Number()` has rounded them: for example, `--max-budget-usd 9007199254740990.5` parses to `9007199254740990`. That rounded value is passed to the downstream `getTotalCost() >= maxBudgetUsd` guard, so the enforced spending cap differs from the one the user supplied—the exact case the new safe-range rationale says it prevents. Validate the decimal representation/precision (or reject non-representable decimal caps) and add regressions for the boundary decimals.

**Comment by 0xfandom** [comment]
Fixed. The MAX_SAFE_INTEGER bound only caught whole-number overflow; a fractional value just under it still rounded (`9007199254740990.5` → `...990`), so the enforced cap drifted from the input. A double round-trips ~15 significant digits, so a fractional input carrying more is now rejected (gated on a decimal point so the 16-digit boundary integer still parses). Added boundary-decimal regressions.

**Review by jatmn — CHANGES_REQUESTED** [review]
I found issues that need to be addressed before this is ready.

## Findings

- [P2] Scientific notation still bypasses the budget precision guard
  `src/utils/cli/numericArgs.ts:63`
  The new `significantDigitCount` check only runs when the raw string contains a literal `.`, so exponent-form inputs skip it entirely. Decimal literals the guard rejects are still accepted via scientific notation with no `.` in the string — for example `9007199254740990.5` is rejected, but `90071992547409905e-1` is accepted (both coerce to the same double, `9007199254740990`), and `1234567890.1234567` is rejected while `12345678901234567e-7` is accepted to the same value. That leaves the precision policy inconsistent: inputs I asked this guard to block can be rewritten in scientific form and pass. Realistic budgets like `0.25` are fine, but the high-precision boundary cases this guard was added for remain bypassable on head — the same gap from my last review and CodeRabbit's inline comment. Please apply the precision check to exponent notation as well (treat negative exponents as fractional input, or round-trip/reject all spellings consistently) and add the `90071992547409905e-1` regression CodeRabbit requested.

- [P3] `MAX_SAFE_INTEGER` boundary rejects the `.0` spelling
  `src/utils/cli/numericArgs.ts:63`
  `--max-budget-usd 9007199254740991` is accepted, but `--max-budget-usd 9007199254740991.0` throws the precision error even though both denote the same representable boundary integer — contrary to the comment at lines 48–50 that the decimal gate should not reject the boundary itself. Low practical impact at this magnitude, but generated flags that append `.0` will fail startup while the bare integer succeeds. I'd normalize or exempt trailing-zero fractional parts that do not change the parsed value while you're in this code.

**Comment by 0xfandom** [comment]
Fixed. The precision guard now decides integrality from the digit string and the exponent-adjusted decimal position instead of a literal `.`, so `90071992547409905e-1` is rejected like its `9007199254740990.5` spelling, and `9007199254740991.0` (the boundary integer) is accepted again. Added the `90071992547409905e-1` regression plus the `.0`/scientific-notation cases.

**Review by jatmn — APPROVED** [review]
@kevincodex1 LGTM

**Comment by jatmn** [comment]
@kevincodex1 please merge

**Comment by 0xfandom** [comment]
Closing this — main has since landed an equivalent `--max-turns` validator (`parseMaxTurnsCommanderArgument` / `parseMaxTurnsCli` in `src/utils/replMaxTurns.ts`), which rejects non-integer and negative values through the Commander arg parser. That covers the NaN/garbage case this PR was guarding against, so this is now redundant. Thanks for the review here.
